src/Processors/Formats/ISchemaWriter.h

```diff
@@ -0,0 +1,46 @@
+#pragma once
+
+#include "Formats/FormatSettings.h"
+#include "IO/ReadBuffer.h"
+
+namespace DB
+{
+
+struct SchemaValidationError final
+{
+    /// On which line the error happens. If line < 0, it indicates the error is not related to specific line,
+    /// or the validator is unable to provide such information.
+    int line;
+    /// At which column the error happens. If col < 0, it indicates the error is not related to specific column,
+    /// or the validator is unable to provide such information.
+    int col;
+    /// Error message
+    String error;
+};
+
+using SchemaValidationErrors = std::vector<SchemaValidationError>;
+
+class IExternalSchemaWriter
+{
+public:
+    IExternalSchemaWriter(std::string_view & schema_body_, const FormatSettings & settings_)
+        : schema_body(schema_body_)
+        , settings(settings_)
+        {}
+
+    virtual ~IExternalSchemaWriter() = default;
+
+    /// Validates the schema input.
+    virtual SchemaValidationErrors validate() = 0;
+
+    /// Persistents the schema.
+    /// If the schema already exists, and replace_if_exist is false, it returns false.
+    /// Otherwise it returns true. Throws exceptions if it fails to write.
+    virtual bool write(bool replace_if_exist) = 0;
+
+protected:
+    std::string_view schema_body;
```

Comment:
save a string view, it's risk
if we can assume it always exists, use `const String &` is a better way

Reply:
The Writer is deisnged to not to own the schema body, so yes, it assumes it out-lives the body string ( I belive this contract is clear because the constructor function takes a view as a parameter ).

I can switch it to a `const String &` if that's neccessary. But could you elaborate a bit on why `const String &` is better in this case, please?

Reply:
Because it can display external references more intuitively,
view also is good.